From 11c0bb6f30703fce6c42d8790ff9ae8ecaf42be1 Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Fri, 30 Aug 2019 16:15:39 +0530
Subject: [PATCH 335/767] octeontx2-pf: Add barrier to sync interface status

commit 446d57d2e083ee57defc821d23c52687cc880d75 from
git@git.assembla.com:cavium/WindRiver.linux.git

Ensure that intf_down flag is updated by using a barrier
before enabling traffic.

Change-Id: I61e003346030b69278d56b4d2a7bba5891470284
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/15093
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c | 6 ++++--
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c      | 2 ++
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
index 7e792f2d7178..eec6c13122bb 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
@@ -457,8 +457,10 @@ static int otx2_set_coalesce(struct net_device *netdev,
 		pfvf->cq_ecount_wait = min_t(u16, ec->rx_max_coalesced_frames,
 					     ec->tx_max_coalesced_frames);
 
-	for (qidx = 0; qidx < pfvf->hw.cint_cnt; qidx++)
-		otx2_config_irq_coalescing(pfvf, qidx);
+	if (netif_running(netdev)) {
+		for (qidx = 0; qidx < pfvf->hw.cint_cnt; qidx++)
+			otx2_config_irq_coalescing(pfvf, qidx);
+	}
 
 	return 0;
 }
diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
index 283c71311e4b..7c9db393c713 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
@@ -1576,6 +1576,8 @@ int otx2_open(struct net_device *netdev)
 	otx2_set_cints_affinity(pf);
 
 	pf->intf_down = false;
+	/* 'intf_down' may be checked on any cpu */
+	smp_wmb();
 
 	err = otx2_rxtx_enable(pf, true);
 	if (err)
-- 
2.31.1

